### PR TITLE
fix(tile-converter): support specific gltf structure

### PR DESCRIPTION
**Pull request by belom88**

### CODE REVIEW

**Review by dryabinin-actionengine — APPROVED** [review]

**Review by Tamrat-B — APPROVED** [review]
Looks good! sample data passes. Pls merge soon. thx!

File: modules/tile-converter/src/i3s-converter/helpers/geometry-converter.js [hunk 1]
```diff
@@ -158,7 +161,7 @@ function convertAttributes(tileContent) {
     });
   }
 
-  const nodes = tileContent.gltf.scene.nodes;
+  const nodes = (tileContent.gltf.scene || tileContent.gltf.scenes?.[0] || tileContent.gltf).nodes;
```
Comment (ibgreen) [inline modules/tile-converter/src/i3s-converter/helpers/geometry-converter.js:164]: 1. This doesn't look correct. IIRC `gltf.scene` is an index into the scene array.
2. Easier to read if we split such logic on multiple lines
3. Good if we can add a link to an "offending" tileset in the comment.

```suggestion
  const sceneIndex = tileContent.gltf.scene || 0;
  const scene = tileContents.gltf.scenes?.[sceneIndex];
  // Note: Some tilesets are not glTF conformant and do not have a `scenes` array, just `nodes`
  const nodes = scene ? scene.nodes : tileContents.gltf.nodes;
```

File: modules/tile-converter/src/i3s-converter/helpers/geometry-converter.js [hunk 2]
```diff
@@ -40,8 +40,8 @@ export default async function convertB3dmToI3sGeometry(
 
   const result = [];
   let nodesCounter = nodeId;
-  for (let i = 0; i < (tileContent.gltf.materials.length || 1); i++) {
-    const sourceMaterial = tileContent.gltf.materials[i] || {id: 'default'};
+  for (let i = 0; i < (tileContent.gltf.materials?.length || 1); i++) {
+    const sourceMaterial = tileContent.gltf.materials?.[i] || {id: 'default'};
     if (!convertedAttributesMap.has(sourceMaterial.id)) {
```
Comment (ibgreen) [inline modules/tile-converter/src/i3s-converter/helpers/geometry-converter.js:45]: There is a lot of conditional code here. Logic would be easier to read if separated on multiple lines.
```suggestion
    let {materials = []} = tileContent.gltf;
    if (materials.length === 0) {
      materials.push({id: 'default'};
    }
    
    for (const sourceMaterial of  materials) {
    const sourceMaterial = materials[i];
    if (!convertedAttributesMap.has(sourceMaterial.id)) {
```

Also why do we use a `continue`? Can't we just move the code below into a non-negated if statement?

**Review by ibgreen — APPROVED** [review]

File: modules/tile-converter/src/i3s-converter/helpers/geometry-converter.js (hunk 1 above)
Comment (belom88) [reply modules/tile-converter/src/i3s-converter/helpers/geometry-converter.js:164]: Our GLTF postprocessing replaces this index to real object

File: modules/tile-converter/src/i3s-converter/helpers/geometry-converter.js (hunk 2 above)
Comment (belom88) [reply modules/tile-converter/src/i3s-converter/helpers/geometry-converter.js:45]: > Also why do we use a continue? Can't we just move the code below into a non-negated if statement?

I remember a case when there is a material that is not linked with some primitive. We have to skip them.

Comment (belom88) [reply modules/tile-converter/src/i3s-converter/helpers/geometry-converter.js]: > There is a lot of conditional code here. Logic would be easier to read if separated on multiple lines.

Done

**Pull request merged**